Sources/NIOHTTP2/HTTP2CommonInboundStreamMultiplexer.swift

```diff
@@ -213,6 +213,30 @@ extension HTTP2CommonInboundStreamMultiplexer {
         self.streamChannelContinuation?.finish()
     }
 
+    internal func propagateErrorCaught(_ error: any Error) {
+        for channel in self.streams.values {
+            channel.receiveStreamClosed(nil)
+        }
+
+        for channel in self._pendingStreams.values {
+            channel.receiveStreamClosed(nil)
+        }
+
+        self.streamChannelContinuation?.finish(throwing: error)
+    }
+
+    internal func propagateHandlerRemoved() {
+        for channel in self.streams.values {
+            channel.receiveStreamClosed(nil)
+        }
+
+        for channel in self._pendingStreams.values {
+            channel.receiveStreamClosed(nil)
+        }
+
+        self.streamChannelContinuation?.finish()
+    }
```

Comment:
I don't think we should be closing the streams in these cases, we should just finish the continuation

Reply:
For `errorCaught` I think you're right (although I'm not sure what it would mean if there was a connection error fired down the pipeline but we keep the streams opened?); however if the handler was removed, wouldn't it make sense to close the streams? Is there anything we can do with them if the connection has died?

Reply:
In the `HTTP2StreamMultiplexer`, handler removed calls `clearDidReadChannels` in the common (i.e. this) multiplexer: https://github.com/apple/swift-nio-http2/blob/91136a6604e58e4f1e6cab24086b876fe01bc24c/Sources/NIOHTTP2/HTTP2StreamMultiplexer.swift#L43-L46

I don't think we should have divergent behaviour between the two. When we reach `handlerRemoved` the streams will have already been closed.

---

For error caught, the `HTTP2StreamMultiplexer` only propagates stream errors to the child channels, they then get to decide how to act on those errors. Any connection level errors aren't propagated to streams and go down the channel pipeline and another handler decides how to act on it (usually by closing the connection, which will result in the streams being closed), but we should avoid short circuiting that because not all errors result in the connection being closed.

Reply:
Thanks George, that's clear. I'll make the change.